### ISSUE TITLE
Automation: "set permission" granularity

### ISSUE THREAD

**Issue by youennf** [w3c/permissions#181]
I looked at the spec in the light of WebKit implementation of a preliminary "set permission" automation API for getUserMedia.

In WebKit, getUserMedia prompt is based on the top level origin only as this is the only thing we can expect a user to reason about.
If a third-party context calls getUserMedia, the decision is still made based on the top level origin.

The current WebKit implementation of "set permission" is using a single value for all browsing contexts.
The current spec allows granularity at the context level and it is not clear to me what the use cases for such a granularity are.

I wonder whether a single value per top origin might be granular enough.
For getUserMedia at least, that would align somehow with how persistent permissions work in Safari.

**Comment by youennf** [comment]
cc @jugglinmike, @agouaillard-cosmo

**Comment by jugglinmike** [comment]
For context, here's the reference to the "current browser context":

> 5. Let settings be the environment settings object of the current browsing
>    context’s active document. 

This is fine-grained, but we designed it that way for completeness: a web application may include an iframe which requests permissions, and we wanted the "Set Permissions" command to support such use cases.

It sounds like Safari doesn't recognize the distinction. I may be mistaken, but I believe the current text allows such a simplification:

> 10. [...]
>     1. [...]
>        1. Interpret parameters.state as if it were the result of an
>           invocation of permission state for typedDescriptor with the
>           argument target made at this moment. 

That step is intentionally vague, written to allow exactly the kind of variance that Safari exhibits. If this text were implemented in Safari, wouldn't a call to "Set Permissions" within a nested browsing context trigger the "single value for all browsing contexts" behavior that you've described?

**Comment by jugglinmike** [comment]
Negative, but I know people who do. @kereliuk and @andreastt helped out with the text, and they work on ChromeDriver and GeckoDriver respectively. @thejohnjansen can probably tell us where Edge is at.

Would you folks mind letting us know where your WebDriver implementation stands on [the "Set Permissions" extension command](https://w3c.github.io/permissions/#set-permission-command)?

**Comment by andreastt** [comment]
The fact that something is in a W3C standard does not imply a browser implements it. Gecko certainly doesn’t at the moment, but I think WPT tests for this might help incentivise progress.

**Comment by foolip** [comment]
@kereliuk, is there a tracking bug for implementing this in ChromeDriver? And in testdriver.js?

If https://w3c.github.io/permissions/#set-permission-command has enough leeway to allow for the behavior that Safari has, then what we could do in testdriver.js is just to not allow use of the API except in top-level frames, so that tests can't depend on such differences. That'd be enough to unblock almost all getUserMedia and WebRTC tests, I'd bet.

**Comment by youennf** [comment]
> If https://w3c.github.io/permissions/#set-permission-command has enough leeway to allow for the behavior that Safari has, then what we could do in testdriver.js is just to not allow use of the API except in top-level frames, so that tests can't depend on such differences. That'd be enough to unblock almost all getUserMedia and WebRTC tests, I'd bet.

Sounds like a good step forward.

> That step is intentionally vague, written to allow exactly the kind of variance that Safari exhibits. If this text were implemented in Safari, wouldn't a call to "Set Permissions" within a nested browsing context trigger the "single value for all browsing contexts" behavior that you've described?

Probably.
The issue is that if the spec allows very different behaviors, the interop story might not be very good.
It would be good to understand the scenarios that require this highly flexible approach.
Maybe we can do some simplification here which might actually help both implementors and users of this API.

**Comment by youennf** [comment]
cc @burg since he has more background on this area

**Comment by jan-ivar** [comment]
> The current spec allows granularity at the context level and it is not clear to me what the use cases for such a granularity are.

@youennf I agree. User permission management is an area of differentiation between browsers. For `query`, context level let us avoid making inferences about the various permission models of browsers. But specifying a per-context *set* method would either expose browser-specific behavior, or require us to specify exactly what it should affect. I don't think we should specify a permission model.

@jugglinmike We only want to test standard stuff. If we automate global blocking and granting (which all browsers support AFAIK), is that not sufficient to test our failure paths? Do we need anything else?

**Comment by jugglinmike** [comment]
> We only want to test standard stuff.

I agree. [The specification has text that allows the user to control behavior at the granularity of a realm](https://w3c.github.io/permissions/#prompt-the-user-to-choose):

> 4. If the user chose an option, return it; otherwise return "denied". If the
>    user’s interaction indicates they intend this choice to apply to other
>    realms, then treat this this as new information about the user’s intent
>    for other realms with the same origin. 

Earlier in this dicussion, I framed this in terms of "completeness"; this is what I had in mind ([more background here](https://github.com/w3c/permissions/pull/151#discussion_r124188076)). If a web developer can write application code that reacts to this behavior, then the automation API ought to allow them to test that code.

It seems as though an objection to this level of granularity in the automation API would apply to the core specification. Do you think that spec text should be made more precise?

Maybe what's needed is a way for UAs to signal that they do not support this optional behavior. Would a change like this make sense?

```diff
-7. If parameters.oneRealm is true, let targets be a list whose sole member is
-   settings. 
+7. If parameters.oneRealm is true, then
+   1. If the permission state cannot be set for a single realm for any
+      implementation-defined reason, return a WebDriver error with WebDriver
+      error code "unsupported operation".
+   2. Otherwise, let targets be a list whose sole member is settings. 
 8. Otherwise, let targets be a list containing all environment settings
    objects whose origin is the same as the origin of settings. 
```

**Comment by youennf** [comment]
Specification should hopefully help getting interoperability between browsers but, as @jan-ivar said, I am not sure how this is feasible in this specific case.
To make progress on the spec, two interoperable implementations of this feature might be required which might be another issue.
At least for WebKit, implementing this might need to violate some internal layering in the WebKit implementation.

The idea to enable testing of browser-specific behaviors makes sense to me.
WebDriver browser-specific extension APIs might better fit the need here.

**Comment by jugglinmike** [comment]
> I think WPT tests for this might help incentivise progress.

@andreastt I've submitted an initial set here: https://github.com/web-platform-tests/wpt/pull/12118

**Comment by jan-ivar** [comment]
> The specification has text that allows the user to control behavior at the granularity of a realm

@jugglinmike, that's not how I read it. It allows user *agents* that want to implement such granularity to do so, but the "other realms" text is carte blanche as far as granularity goes.

Specifically, I read no requirement here that browsers MUST offer its users discrete permission at a per-realm level.

This note in the spec seems to support this:

  *"This is intentionally vague about the details of the permission UI and how the UA infers user intent. UAs should be able to explore lots of UI within this framework."*
